### PR TITLE
Remove secure proxy option, replace Override.Secure in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR makes the following updates to Ice for C#:
- it removes the "secure" proxy option. Ice still parses and unmarshals successfully a proxy with this option, however when it marshals or stringify such a proxy, the secure option is always set to false
- likewise, it removes `Ice.Override.Secure` (there was no `Ice.Default.Secure`)
- it replaces the proxy option `PreferSecure` (and associated default `Ice.Default.PreferSecure`), with the proxy option `PreferNonSecure` (and associated default `Ice.Default.PreferNonSecure`). The default for `PreferNonSecure` is 1 (true) for now, but should be changed to 0 (false) in the near future
The semantics of `PreferNonSecure` is _not_ the opposite of the old `PreferSecure`: PreferNonSecure=true is equivalent to PreferSecure=false i.e. Ice first attempts to establish a non-secure connection and falls back to a secure connection. PreferSecure=true (in 3.7) is just a preference and falls back to a non-secure connection. That's not the case with PreferNonSecure=false, which does not fall back to a non-secure connection.
- it adds a new configuration property `Ice.AcceptNonSecureConnections`, which controls whether object adapters accept incoming non-secure connections. Currently only the property is implemented.

### CODE REVIEW

File: csharp/src/Ice/Communicator.cs [hunk 1]
```diff
@@ -44,11 +44,13 @@ private class ObserverUpdater : Instrumentation.IObserverUpdater
...
         }
 
-        /// <summary>
-        /// Get the logger for this communicator.
-        /// </summary>
-        /// <returns>This communicator's logger.</returns>
-        public ILogger Logger { get; internal set; }
+         /// <summary>Indicates whether or not this communicator allows non-secure connections. When false, only
+         /// secure connections are allowed; when true, both secure and non-secure connections are allowed. This
+         /// property applies to both incoming and outgoing connections are corresponds to the Ice.AllowNonSecure
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs:49]: ` connections are corresponds to the Ice.AllowNonSecure` typo there?

File: csharp/src/Ice/IObjectPrx.cs [hunk 2]
```diff
@@ -177,128 +177,85 @@ public string IceId(IReadOnlyDictionary<string, string>? context = null)
...
-        /// </summary>
-        /// <returns>The endpoints used by this proxy.</returns>
+        /// <summary>The endpoints of this proxy. A proxy with a non-empty endpoint list is a direct proxy.</summary>
         public IEndpoint[] Endpoints => (IEndpoint[])IceReference.Endpoints.Clone();
 
-        /// <summary>
-        /// Returns the locator cache timeout of this proxy.
-        /// </summary>
-        /// <returns>The locator cache timeout value (in seconds).</returns>
+        /// <summary>The locator cache timout of this proxy, in seconds.</summary>
```
Comment (pepone) [inline csharp/src/Ice/IObjectPrx.cs:197]:
```suggestion
        /// <summary>The locator cache timeout of this proxy, in seconds.</summary>
```

File: csharp/src/Ice/Reference.cs [hunk 3]
```diff
@@ -1311,7 +1273,7 @@ public void SetException(System.Exception ex)
         // TODO: refactor this class
         private sealed class EndpointComparator : IComparer<Endpoint>
```
Comment (pepone) [inline csharp/src/Ice/Reference.cs:1274]: Seems this class is no longer used

File: csharp/src/Ice/Communicator.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator.cs:49]: Should be "and". Fixed.

File: csharp/src/Ice/IObjectPrx.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/IObjectPrx.cs:197]: Fixed.

File: csharp/src/Ice/Reference.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/Reference.cs:1274]: Removed.

File: csharp/src/Ice/Communicator.cs [hunk 4]
```diff
@@ -943,12 +952,12 @@ public Reference CreateReference(string s, string? propertyPrefix = null)
 
                     case 's':
                         {
+                            // we ignore this secure option, just parse it
```
Comment (bentoi) [inline csharp/src/Ice/Communicator.cs:955]: Shouldn't we print a warning to explain that it's no longer supported?

File: csharp/src/Ice/OpaqueEndpointI.cs [hunk 5]
```diff
@@ -61,7 +61,8 @@ public override void StreamWrite(Ice.OutputStream s)
         public override string ToString()
         {
             string val = System.Convert.ToBase64String(_rawBytes);
-            return "opaque -t " + _type + " -e " + _rawEncoding.ToString() + " -v " + val;
+            short typeNum = (short)_type;
+            return $"opaque -t {typeNum.ToString(CultureInfo.InvariantCulture)} -e {_rawEncoding} -v {val}";
```
Comment (bentoi) [inline csharp/src/Ice/OpaqueEndpointI.cs:65]: If integer formatting is culture dependent, we need to fix other ToString methods.

Comment (pepone) [reply csharp/src/Ice/OpaqueEndpointI.cs:65]: Without `InvariantCulture` you can end up with decimal separators in the string.

File: csharp/src/Ice/Communicator.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator.cs]: Added warning.

**Pull request merged**